### DISCUSSION TITLE
finial project 使用 React Next.js 在 build 時遇到問題 😂

### DISCUSSION THREAD

**Discussion by roroiii** [Lidemy/forum#235] [前端相關問題]
標題：[FE201] 有關第六個作業

### 碰到的問題
跟 next.js getStaticPaths 有關的問題，有問題的程式碼在此：
https://github.com/david-christian/cakeShoping/blob/backstage/cakeshoping/pages/admin/products/%5Bid%5D.js

#### 問題一、 next.js 無法 build 專案，`getStaticPaths` 回報錯誤 `type: 'TypeError'`
![](https://i.imgur.com/5kmkhdQ.png)

#### 問題二、 在本地端  `yarn dev` 開發時會有錯誤訊息出現，但是可以正常開發，api 傳回來的資料都正確
![](https://i.imgur.com/m3NSe2s.png)
![](https://i.imgur.com/TYCxNII.png)

### 預期的行為
應該要可以 `yarn build` 起來專案，console 沒有錯誤訊息
 
#### 問題一
搜尋「getStaticPaths type: 'TypeError'」沒有找到合適的解答
把 ids 跟 paths 放到 getStaticProps() 裡面嘗試在頁面 console.log() 出來查看，回傳的型態看起來是正常的
![](https://i.imgur.com/ufPk1Xd.png)

#### 問題二
搜尋「react-dom.development.js?61bb:67 Warning: Prop `className` did not match. Server:」
找到解決方案https://github.com/vercel/next.js/issues/7322
嘗試過 
1. `.babelrcs` 跟 `babel.config.json` 都加入
```js
"plugins": [
    ["styled-components", { "ssr": true, "displayName": true }]
  ]
```
2. 刪除 .next 資料夾重新跑一次專案
3. 安裝 babel-plugin-styled-components

以上三種方法都無法解決問題二

### 相關程式碼
https://github.com/david-christian/cakeShoping/blob/backstage/cakeshoping/pages/admin/products/%5Bid%5D.js
```js
export const getStaticProps = async (content) => {
  const { data: product } = await getProduct(content.params.id);
  const { data: photos } = await getPhoto(content.params.id);

  return {
    props: {
      product: product.result[0],
      photos: photos.result,
    },
  };
};

export const getStaticPaths = async () => {
  const { data: res } = await getProducts();
  const products = res.result;

  const ids = products.map((product) => product.id);
  const paths = ids.map((id) => ({ params: { id: id.toString() } }));

  return {
    paths,
    fallback: false,
  };
};
```

### 補充說明
不知道是哪裡出錯惹，雖然不影響開發但還是想解決這個問題，還是因為使用 [id].js 當作檔名的關係？
看了官網給的 [getStaticPaths](https://nextjs.org/docs/basic-features/data-fetching) 範例也找不到問題出在哪QQ
希望老師助教同學們可以給我個明燈，謝謝

**Comment by aszx87410** [comment]
問題二只是 warning 可以先不用管他，問題一的話你搞錯方向了，沒有看清楚錯誤訊息
錯誤重點不是「TypeError」，而是在更上面的：「TypeError: cannot read property 'map' of undefined」
從這個就可以很明確知道問題是因為在 undefined 上面用了 map
程式碼裡面有用到 map 的就那兩行，你把 products 跟 ids 加上 console.log 然後再 build 一次，應該就能繼續 debug 了

**Reply by roroiii** [reply]
後來改用 fetch 型態就正確了，可能是 axios 回傳的型態不符合， products 跟 ids 加上 console.log 資料都是對的，回傳都不是 undefioed 所以覺得很奇怪

```js
export const getStaticPaths = async () => {
  const res = await fetch(`${server}/product`);
  const products = await res.json();

  const paths = products.result.map((post) => ({
    params: { id: post.id.toString() },
  }));

  return {
    paths,
    fallback: false,
  };
};
```
`yarn build` 的腳本是 `next build && next export`
問題一修好之後噴了很多新的錯誤😂
 
好像next build 起來有很多細節要調整才不會有問題，但是在 dev 模式的時候都沒有這些錯誤
打算最後開發完要 build 的時候再一起全部 debug
不然會一直卡在這些錯誤裡 

謝謝老師給的方向😉
